### PR TITLE
added puppetrun over ssh support

### PR DESCRIPTION
**Pull request by cyberkov**
Hello!

I wrote a puppetrun over ssh extension. 
Unfortuantely `( Process.detach(c.pid); return 0 ) unless wait` doesn't seem to be very ruby-like...
Maybe someone can help me out with that one :-)

Cheers
Hannes

### CODE REVIEW

**Comment by lzap** [comment]
You know, maybe it's fine. I am not great rubyist :-)

Thanks for the patch!

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

**Comment by ohadlevy** [comment]
[test]

**Comment by ares** [comment]
How about

``` ruby
return 0 if Process.detach(c.pid) && !wait
```

or just simply (since I'm not sure about return value of detach)

``` ruby
unless wait
  Process.detach(c.pid)
  return 0
end
```

**Comment by cyberkov** [comment]
Thank you :) 
@ares I think go with the "unless wait" option just to be on the safe side :)

I also managed to put in some tests for the module but i keep getting ssh errors at the end. I'm not sure how I could suppress those.

```
Finished tests in 0.954535s, 143.5254 tests/s, 239.9075 assertions/s.

137 tests, 229 assertions, 0 failures, 0 errors, 0 skips
ssh: Could not resolve hostname host1: Name or service not known
ssh: Could not resolve hostname host2: Name or service not known
```

**Comment by lzap** [comment]
I can confirm, investigating.

File: test/puppetssh_test.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    assert !@puppetssh.run
+  end
+
+  def test_missing_sudo_and_not_needed
+    SETTINGS.stubs(:puppetssh_sudo).returns(false)
+    @puppetssh.stubs(:which).with("sudo", anything).returns(false)
+    @puppetssh.stubs(:which).with("ssh", anything).returns("/usr/bin/ssh")
+
+    assert @puppetssh.run
```
Comment (lzap) [inline test/puppetssh_test.rb]: Okay you are missing this

```
@puppetssh.stubs(:shell_command).returns(true)
```

before the assert in this test. Thats it.

Looks good to me.

**Comment by skottler** [comment]
Thanks for your PR @cyberkov. So I think there are a few things this needs before it can get merged:
1. The ability to specify a key. `ssh-agent` likely won't always have the right key available, which will break in the current design.
2. Input needs to be properly escaped.

I'm not particularly keen on having the proxy be able to SSH to all the hosts it manages. What's the use case for this beyond just simplicity?

File: lib/proxy/puppet/puppetssh.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      cmd.push(which("ssh", "/usr/bin"))
+
+      if cmd.include?(false)
+        logger.warn "sudo or the ssh binary is missing."
+        return false
+      end
+
+      ssh_command = (SETTINGS.puppetssh_command || '/usr/bin/puppet agent --onetime --no-usecacheonfailure')
+      nodes.each do |node|
+        shell_command(cmd + [ node, ssh_command ], false )
```
Comment (skottler) [inline lib/proxy/puppet/puppetssh.rb]: `Proxy::Util.escape_for_shell` should get used here. It's a 1.8 + 1.9 compatible version of the escape methods that are available in 1.9.

**Comment by ohadlevy** [comment]
note that we also have something similar in code (thanks to @lzap who wrote it a long time ago) 
https://github.com/theforeman/smart-proxy/blob/develop/lib/proxy/util.rb#L15

maybe we can reuse?

**Comment by lzap** [comment]
Hmmm this had to be the other lzap, the smarter one :-)

Indeed, we might want to refactor shell_command to use this utility spawner, which improves logging a bit.

This might be out of scope of this effort tho. Would be a bonus.

**Comment by lzap** [comment]
@skottler I think both your points are resolved now! LGTM [test]

on my system:

```
140 tests, 238 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by ohadlevy** [comment]
I think it make sense to use the CommandTask class here, to have a consistent usage across the app, @cyberkov if you can't refactor it, let me know and I'll fix it myself.

**Comment by ohadlevy** [comment]
thanks! merged with minor changes as 13ed471 for 1.3!

**Pull request closed without merging**